src/foam/core/AbstractAgency.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'AbstractAgency',
+  abstract: true,
+  implements: [ 'foam.core.Agency' ],
+
+  javaImports: [
+    'foam.nanos.logger.Loggers',
+    'java.util.Timer',
+    'java.util.TimerTask',
+    'java.util.concurrent.ConcurrentHashMap'
+  ],
+
+  javaCode: `
+    protected static final Timer TIMER = new Timer();
+    protected static final ConcurrentHashMap<String, TimerTask> TASK_QUEUE = new ConcurrentHashMap<>();
+
+    public void schedule(X x, ContextAgent agent, String key, long delay) {
+      if ( delay <= 0 ) {
+        submit(x, agent, key);
+        return;
+      }
+
+      var task = TASK_QUEUE.get(key);
+
+      // Cancel the task if it's not yet executed and being re-scheduled
+      if ( task != null
+        && task.scheduledExecutionTime() - System.currentTimeMillis() > 0
+      ) {
+        task.cancel();
```

Comment:
there is a race condition here .. in the delay for the old task and the cancellation of it by the new task coming in .. why not keep the old task instead of canceling it?